src/journal-remote/journal-header-util.c

```diff
@@ -0,0 +1,96 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "escape.h"
+#include "journal-header-util.h"
+#include "string-util.h"
+#include "strv.h"
+
+/* HTTP header name can contains:
+- Alphanumeric characters: a-z, A-Z, and 0-9
+- The following special characters: - and _ */
+#define VALID_HEADER_NAME_CHARS \
+        DIGITS LETTERS "_-"
+
+#define VALID_HEADER_NAME_LENGTH 40
+
+#define VALID_HEADER_VALUE_CHARS \
+        DIGITS LETTERS "_ :;.,\\/'\"?!(){}[]@<>=-+*#$&`|~^%"
+
+static bool header_name_is_valid(const char *e) {
+        if (!e)
+                return false;
+
+        if (strlen(e) > VALID_HEADER_NAME_LENGTH)
+                return false;
+
+        return in_charset(e, VALID_HEADER_NAME_CHARS);
+}
+
+static bool header_value_is_valid(const char *e) {
+        if (!e)
+                return false;
+
+        return in_charset(e, VALID_HEADER_VALUE_CHARS);
+}
+
+int header_put(OrderedHashmap **headers, char *name, char *value) {
+        int r;
+
+        assert(headers);
+
+        if (!header_value_is_valid(value))
+                return 0;
+
+        if (!header_name_is_valid(name))
+                return 0;
+
+        r = string_strv_ordered_hashmap_put(headers, name, value);
+        if (r < 0)
+                return log_oom();
+
+        return r;
+}
```

Comment:
this function is so low-level. i think it's better to do error logging in caller, not in here